### PR TITLE
Chima: Setup linting with eslint and prettier

### PR DESCRIPTION
**Pull request by chukwuemekachm**
#### What does this PR achieve?
- This PR introduces the required changes to setup linting on the project with eslint and prettier. It also lints the current files and resolves all linting issues.

#### Any background context?
- N/A

#### Manual Testing Steps
- Run `yarn` to install dependencies
- Run `yarn lint` to test linting and view results
- Make any commit and see the commit hooks in action

#### Screenshots:
- N/A

#### Relevant Trello Tickets
- [Ticket #42](https://trello.com/c/mQWPMBEA)

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+module.exports = {
+  parser: "@typescript-eslint/parser",
+  extends: [
+    "plugin:react/recommended",
+    "plugin:@typescript-eslint/recommended",
+    "prettier/@typescript-eslint",
+    "plugin:prettier/recommended"
```
Comment (chidioguejiofor) [inline .eslintrc.js:7]: I thought the original plan was to use Airbnb style guide. Is this `plugin:prettier/recommended` generally the same

File: package.json [hunk 2]
```diff
@@ -10,7 +10,19 @@
     "build:prod": "rm -rf dist && webpack --env.mode production",
     "prestart": "yarn build:prod",
     "start": "node server.js",
-    "dev": "yarn build:dev && rm -rf dist && webpack-dev-server --env.mode development --open --hot"
+    "dev": "yarn build:dev && rm -rf dist && webpack-dev-server --env.mode development --open --hot",
+    "lint": "rm -rf dist && tsc --noEmit && eslint '*/**/*.{js,ts,tsx}' --quiet --fix"
+  },
+  "husky": {
+    "hooks": {
+      "pre-commit": "tsc --noEmit && lint-staged"
```
Comment (chidioguejiofor) [inline package.json:19]: Brilliant!

File: package.json [hunk 3]
```diff
@@ -10,7 +10,19 @@
...
+    }
+  },
+  "lint-staged": {
+    "*.{js,ts,tsx}": [
+      "yarn lint",
+      "git add"
```
Comment (chidioguejiofor) [inline package.json:25]: Nice detail here

File: src/components/containers/AuthenticationContainer/index.ts [hunk 4]
```diff
@@ -87,32 +69,44 @@ export default function withAuthenticationContainer(WrappedComponent) {
       return signUp({ ...values, redirectURL: SIGN_UP_REDIRECT_URL });
     }
 
-    async function handleAuthOperation(schemaKey, validFuncName) {
+    async function handleAuthOperation(schemaKey: string, validFuncName: string): Promise<void> {
```
Comment (chidioguejiofor) [inline src/components/containers/AuthenticationContainer/index.ts:72]: We're now `typing` the script! 💪🏻

File: src/components/pages/Authentication/SignUp/index.tsx [hunk 5]
```diff
@@ -1,20 +1,21 @@
...
 
 import SignUpForm from './SignUpForm';
 import withAuthenticationContainer, { AuthenticationProps } from 'components/containers/AuthenticationContainer';
-import __spacing from 'settings/__spacing';
 import { signUp } from 'store/actions/auth';
 
-function SignUp(props: AuthenticationProps) {
- 
-   return (<SignUpForm {...props} />);
+export interface SignUpProps extends AuthenticationProps {
```
Comment (chidioguejiofor) [inline src/components/pages/Authentication/SignUp/index.tsx]: I think this actually makes a lot of sense but would `LoginProps` not have the same value? Is it possible to `isLoading` to `AuthenticationProps`and then perform this inheritance only for `NewPasswordProps`( which would now have only the `changePassword` key that was added to it)

File: src/components/pages/ConfirmResetPassword/NewPassword/index.tsx [hunk 6]
```diff
@@ -1,17 +1,30 @@
...
 import { BRAND_PRIMARY, BRAND_WHITE, BRAND_PRIMARY_HOVER } from 'settings/__color';
 import { changeUserPassword } from 'store/actions/auth';
-import Button from 'components/ui/Button';
 
-function NewPassword(props: AuthenticationProps) {
-  const { isLoading, handleChange, handleSubmit, handleBlur, values, values: { errors } } = props;
+interface NewPasswordProps extends AuthenticationProps {
+  isLoading: boolean;
+  changePassword: (payload: any) => void;
+}
```
Comment (chidioguejiofor) [inline src/components/pages/ConfirmResetPassword/NewPassword/index.tsx:17]: See comment about inheriting `AuthenticationProps` above

**Review by chidioguejiofor — CHANGES_REQUESTED** [review]
Nice Job @chukwuemekachm . Glad we now have linting in the mix and I also learnt some things while reviewing this one. 
Dropped a few comments that could improve the output though

File: src/components/pages/Authentication/SignUp/index.tsx (hunk 5 above)
Comment (chukwuemekachm) [reply src/components/pages/Authentication/SignUp/index.tsx]: That's right. But I'm also looking at the context in which the `isLoading` prop exists. It is available on those components but not on the Authentication component as well. Adding it there I feel will just add one more prop which will never exist in the Authentication context for the sake of `prop drilling`.

File: .eslintrc.js (hunk 1 above)
Comment (chukwuemekachm) [reply .eslintrc.js:7]: Airbnb is a JS style guide, a lot of the styles don't apply to TypeScript and a lot of TypeScripts best practices is not available on Airbnb. These plugins are the recommended ones for integrating EsLint with TypeScript, React and Prettier as recommended by their authors.

Comment (chidioguejiofor) [reply .eslintrc.js:7]: Nice. Thanks for the clarification.

File: src/components/pages/Authentication/index.tsx [hunk 7]
```diff
@@ -1,48 +1,42 @@
...
 import { connect } from 'react-redux';
+
 import AuthenticationLayout from 'components/layouts/AuthenticationLayout';
 import __spacing from 'settings/__spacing';
 import { fontSizes } from 'settings/__fonts';
 import { GRAY, BLACK, GAINS_BORO } from 'settings/__color';
 import Login from './Login';
 import SignUp from './SignUp';
 
+interface AuthenticationProps {
```
Comment (chidioguejiofor) [inline src/components/pages/Authentication/index.tsx]: I think we might want to use a different name for this variable here as that name had already occurred in other files.

File: src/components/pages/Authentication/SignUp/index.tsx (hunk 5 above)
Comment (chidioguejiofor) [reply src/components/pages/Authentication/SignUp/index.tsx]: > Adding it there I feel will just add one more prop which will never exist in the Authentication context 

If by `Authentication context ` you mean the prop in `pages/Authentication/index.tsx`, that is actually totally different. I dropped a comment to rename of the interface for that component's prop.

If not, then the `AuthenticationProps` that we are talking about here would be used in 4 places:
- SignUp
- Login
- ForgotPassword
- ConfirmPassword

the `isLoading` property in each of these have is used to control the button loader and is actually used in all of them

Comment (chidioguejiofor) [reply src/components/pages/Authentication/SignUp/index.tsx]: > If by Authentication context you mean the prop in `pages/Authentication/index.tsx`, that is actually totally different. I dropped a comment to rename of the interface for that component's prop.

Maybe you can rename this particular interface from  `AuthenticationProps` to `AuthenticationProps`

File: src/components/pages/Authentication/index.tsx (hunk 7 above)
Comment (chidioguejiofor) [reply src/components/pages/Authentication/index.tsx]: See comment on renaming this above ☝️

**Pull request merged**